packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -204,13 +221,54 @@ function createMutationQueue(db: IDBDatabase): void {
     keyPath: DbMutationQueue.keyPath
   });
 
-  db.createObjectStore(DbMutationBatch.store, {
-    keyPath: DbMutationBatch.keyPath as KeyPath
+  const mutationBatchesStore = db.createObjectStore(DbMutationBatch.store, {
+    keyPath: DbMutationBatch.keyPath,
+    autoIncrement: true
   });
+  mutationBatchesStore.createIndex(
+    DbMutationBatch.userMutationsIndex,
+    DbMutationBatch.userMutationsKeyPath,
+    { unique: true }
+  );
 
   db.createObjectStore(DbDocumentMutation.store);
 }
 
+/**
+ * Upgrade function to migrate the 'mutations' store from V1 to V3. Loads
+ * and rewrites all data.
+ */
+function upgradeMutationBatchSchemaAndMigrateData(
+  db: IDBDatabase,
+  txn: SimpleDbTransaction
+): PersistencePromise<void> {
+  const v1MutationsStore = txn.store<DbMutationBatchKey, DbMutationBatch>(
+    DbMutationBatch.store
+  );
+  return v1MutationsStore.loadAll().next(existingMutations => {
+    db.deleteObjectStore(DbMutationBatch.store);
```

Comment:
Hrm...  I just noticed that this doesn't (apparently) happen within the context of the transaction...  which means an inopportune page close or similar could cause data loss.  :-/  We could avoid this by "copying" the mutations into a new store with the new schema rather than delete / rewrite in-place.  But I'm probably okay with the data loss risk. I'd add a comment to make it clear it's intentional though...

Reply:
Oh, nice!